File: src/main/java/priv/bajdcc/LALR1/grammar/tree/ExpValue.java
package priv.bajdcc.LALR1.grammar.tree;

import priv.bajdcc.LALR1.grammar.codegen.ICodegen;
import priv.bajdcc.LALR1.grammar.runtime.RuntimeInst;
import priv.bajdcc.LALR1.grammar.semantic.ISemanticRecorder;
import priv.bajdcc.LALR1.grammar.tree.closure.IClosureScope;
import priv.bajdcc.LALR1.grammar.type.TokenTools;
import priv.bajdcc.util.lexer.token.Token;
import priv.bajdcc.util.lexer.token.TokenType;

/**
 * 【语义分析】基本操作数
 *
 * @author bajdcc
 */
public class ExpValue implements IExp {

	/**
	 * 单词
	 */
	private Token token = null;

	public Token getToken() {
		return token;
	}

	public Token setToken(Token token) {
		return this.token = token;
	}

	@Override
	public boolean isConstant() {
		return token.kToken != TokenType.ID;
	}

	@Override
	public boolean isEnumerable() {
		return false;
	}

	@Override
	public IExp simplify(ISemanticRecorder recorder) {
		return this;
	}

	@Override
	public void analysis(ISemanticRecorder recorder) {

	}

	@Override
	public void genCode(ICodegen codegen) {
		codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(token.object));
		if (token.kToken != TokenType.ID) {
			codegen.genCode(RuntimeInst.iload);
		} else {
			if (TokenTools.isExternalName(token)) {
				codegen.genCode(RuntimeInst.iloadx);
			} else {
				codegen.genCode(RuntimeInst.iloadv);
			}
		}
	}

	@Override
	public String toString() {
		return token.toRealString();
	}

	@Override
	public String print(StringBuilder prefix) {
		return toString();
	}

	@Override
	public void addClosure(IClosureScope scope) {
		if (token.kToken == TokenType.ID) {
			scope.addRef(token.object);
		}
	}

	@Override
	public void setYield() {

	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/tree/ExpValue.kt
package com.bajdcc.LALR1.grammar.tree

import com.bajdcc.LALR1.grammar.codegen.ICodegen
import com.bajdcc.LALR1.grammar.runtime.RuntimeInst
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder
import com.bajdcc.LALR1.grammar.tree.closure.IClosureScope
import com.bajdcc.LALR1.grammar.type.TokenTools
import com.bajdcc.util.lexer.token.Token
import com.bajdcc.util.lexer.token.TokenType

/**
 * 【语义分析】基本操作数
 *
 * @author bajdcc
 */
class ExpValue : IExp {

    /**
     * 单词
     */
    var token: Token? = null

    override fun isConstant(): Boolean {
        return token!!.type !== TokenType.ID
    }

    override fun isEnumerable(): Boolean {
        return false
    }

    override fun simplify(recorder: ISemanticRecorder): IExp {
        return this
    }

    override fun analysis(recorder: ISemanticRecorder) {

    }

    override fun genCode(codegen: ICodegen) {
        codegen.genCode(RuntimeInst.ipush, codegen.genDataRef(token!!.obj!!))
        if (token!!.type !== TokenType.ID) {
            codegen.genCode(RuntimeInst.iload)
        } else {
            if (TokenTools.isExternalName(token!!)) {
                codegen.genCode(RuntimeInst.iloadx)
            } else {
                codegen.genCode(RuntimeInst.iloadv)
            }
        }
    }

    override fun toString(): String {
        return token!!.toRealString()
    }

    override fun print(prefix: StringBuilder): String {
        return toString()
    }

    override fun addClosure(scope: IClosureScope) {
        if (token!!.type === TokenType.ID) {
            scope.addRef(token!!.obj!!)
        }
    }

    override fun setYield() {

    }
}
